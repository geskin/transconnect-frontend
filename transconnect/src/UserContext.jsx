import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

