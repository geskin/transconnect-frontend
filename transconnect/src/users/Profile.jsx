import React, { useContext, useEffect, useState } from "react";
import { useNavigate, useParams, Link } from "react-router-dom";
import { Card, CardContent, Typography, Button, Box, CardActions } from "@mui/material";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import { formatDate } from "../utils/formatDate";

const Profile = ({ logout }) => {
    const { username } = useParams();
    const { currUser } = useContext(UserContext);
    const [user, setUser] = useState(null);
    const navigate = useNavigate();

    if (!username || !currUser) return <Typography>Loading...</Typography>;

    const deleteProfile = async () => {
        if (window.confirm("Are you sure you want to delete your profile? This action cannot be undone.")) {
            try {
                await TransconnectApi.deleteUser(username, user);
                logout();
                navigate("/");
            } catch (err) {
                console.error("Error deleting user:", err);
            }
        }
    };

    useEffect(() => {
        if (!currUser) navigate("/");
        if (username !== currUser.username) {
            const fetchUser = async () => {
                try {
                    const user = await TransconnectApi.getUser(username);
                    setUser(user);
                } catch (err) {
                    console.error("Error fetching user info", err);
                }
            };
            fetchUser();
        }
    }, [username, currUser, navigate]);

    const displayedUser = username === currUser.username ? currUser : user;

    return (
        <Box display="flex" justifyContent="center" mt={4}>
            <Card sx={{ width: 500, p: 3, boxShadow: 3, borderRadius: 2 }}>
                <CardContent>
                    <Typography variant="h4" fontWeight="bold" gutterBottom>
                        {displayedUser?.username}
                    </Typography>
                    {displayedUser?.pronouns && (
                        <Typography variant="subtitle1" color="text.secondary">
                            {displayedUser.pronouns}
                        </Typography>
                    )}
                    {displayedUser?.bio && (
                        <Typography variant="body1" mt={1}>
                            {displayedUser.bio}
                        </Typography>
                    )}
                    {displayedUser?.email && (username === currUser.username || currUser.role === 'ADMIN') && (
                        <Typography variant="body2" color="text.secondary" mt={1}>
                            {displayedUser.email}
                        </Typography>
                    )}
                    <Typography variant="caption" display="block" mt={2}>
                        Created on: {formatDate(displayedUser?.createdAt)}
                    </Typography>
                </CardContent>

                <CardActions>
                    {(username === currUser.username || currUser.role === 'ADMIN') && (
                        <div>
                            <Button
                                variant="contained"
                                color="primary"
                                sx={{ mt: 2 }}
                                component={Link}
                                to={`/users/${username}/edit`}
                            >
                                Edit Profile
                            </Button>
                            <Button
                                variant="contained"
                                color="error"
                                sx={{ mt: 2, ml: 2 }}
                                onClick={deleteProfile}
                            >
                                Delete Profile
                            </Button>
                        </div>
                    )}
                    <Button onClick={() => navigate(`/users/${username}/posts`)} size="small">Posts</Button>
                </CardActions>
            </Card>
        </Box>
    );
};

export default Profile;
