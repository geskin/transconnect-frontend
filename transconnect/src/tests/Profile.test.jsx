import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Profile from "../users/Profile";
import { describe, it, expect, vi, beforeEach } from "vitest";
import UserContext from "../UserContext";
import TransconnectApi from "../api";

vi.mock("../api");

const mockUser = {
    username: "testuser",
    pronouns: "they/them",
    bio: "This is a test bio.",
    email: "test@example.com",
    createdAt: "2025-03-14",
    role: "USER"
};

describe("Profile Component", () => {
    it("renders without crashing", async () => {
        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currUser: mockUser }}>
                    <Profile logout={vi.fn()} />
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => expect(screen.getByText("testuser")).toBeInTheDocument());
    });


    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UserContext.Provider value={{ currUser: mockUser }}>
                    <Profile logout={vi.fn()} />
                </UserContext.Provider>
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("shows user details", async () => {
        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currUser: mockUser }}>
                    <Profile logout={vi.fn()} />
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getByText("testuser")).toBeInTheDocument();
            expect(screen.getByText("they/them")).toBeInTheDocument();
            expect(screen.getByText("This is a test bio.")).toBeInTheDocument();
            expect(screen.getByText("test@example.com")).toBeInTheDocument();
        });
    });

    it("calls delete function on button click", async () => {
        window.confirm = vi.fn(() => true);
        TransconnectApi.deleteUser.mockResolvedValue({});
        const mockLogout = vi.fn();

        const mockUser = { username: "testuser", role: "USER" };

        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currUser: mockUser }}>
                    <Profile logout={mockLogout} />
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => screen.getByText("Delete Profile"));

        fireEvent.click(screen.getByText("Delete Profile"));

        await waitFor(() => expect(mockLogout).toHaveBeenCalled());
    });
});
