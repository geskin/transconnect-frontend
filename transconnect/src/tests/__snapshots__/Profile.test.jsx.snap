// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile Component > matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
  >
    Loading...
  </p>
</DocumentFragment>
`;
